plasma_framework/contracts/src/fast_exits/liquidity.sol

```diff
@@ -0,0 +1,202 @@
+        uint16 firstTransactionOutputIndex = getOutputIndex(_utxoPosInput);
+
+            FungibleTokenOutputModel.Output memory outputFromFirstTransaction
+         = decodedFirstTx.outputs[firstTransactionOutputIndex];
+        address ownerFirstTxOutput = address(
+            uint160(outputFromFirstTransaction.outputGuard)
+        );
+        require(
+            ownerFirstTxOutput == msg.sender,
+            "First Tx Provided was not called by sender"
+        );
+    }
+
+    /**
+     * @notice Verify the First Tx provided is valid
+     * @param _utxoPosInput position of the output that created the inputs for second transaction
+     * @param _rlpInputCreationTx RLP-encoded first transaction that transfers to this contract
+     * @param _inputCreationTxInclusionProof First transactions inclusion proofs
+    */
+    function verifyTxValidity(
+        uint256 _utxoPosInput,
+        bytes memory _rlpInputCreationTx,
+        bytes memory _inputCreationTxInclusionProof
+    ) internal {
+        PosLib.Position memory utxoDecoded = PosLib.decode(_utxoPosInput);
+        utxoDecoded.outputIndex = 0;
+        (bytes32 _root, ) = pf.blocks(utxoDecoded.blockNum);
+        require(_root != bytes32(""), "Failed to get root of the block");
+        bool txExists = Merkle.checkMembership(
+            _rlpInputCreationTx,
+            utxoDecoded.txIndex,
+            _root,
+            _inputCreationTxInclusionProof
+        );
+        require(
+            txExists,
+            "Provided Transaction isn't finalized or doesn't exist"
+        );
+    }
+
+    /**
+     * @notice func that calls omg-contracts to start the exit
+     * @param _utxoPosToExit position of the output which the contract has to exit
+     * @param _rlpOutputTxToContract RLP-encoded transaction that creates the output for the contract
+     * @param _outputTxToContractInclusionProof Second Transaction's inclusion proof
+    */
+    function runExit(
+        uint256 _utxoPosToExit,
+        bytes memory _rlpOutputTxToContract,
+        bytes memory _outputTxToContractInclusionProof
+    ) internal {
+        PaymentStandardExitRouterArgs.StartStandardExitArgs memory s;
+        s.utxoPos = _utxoPosToExit;
+        s.rlpOutputTx = _rlpOutputTxToContract;
+        s.outputTxInclusionProof = _outputTxToContractInclusionProof;
+        peg.startStandardExit.value(msg.value)(s);
+    }
+
+    function getExitId(bytes memory _txBytes, uint16 _outputIndex)
+        internal
+        pure
+        returns (uint160)
+    {
+        uint256 exitId = (uint256(keccak256(_txBytes)) >> 105) |
+            (uint256(_outputIndex) << 152);
+        uint160 croppedExitId = uint160(exitId);
+        require(uint256(croppedExitId) == exitId, "ExitId overflows");
+        return croppedExitId;
+    }
+
+    /**
+     * @dev Get Amount from contract after exit is processed - (to be updated)
+     * @param _exitId The exit id
+    */
+    function getWithdrawl(uint160 _exitId) public {
```

Comment:
can we move this public api in place earlier in the code?

Reply:
yes, definitely once the withdrawal method is finalised